lib/common/digest.c

```diff
@@ -23,22 +23,23 @@
 #define BEST_EFFORT_STATUS 0
 
 /*!
+ * \internal
  * \brief Dump XML in a format used with v1 digests
  *
- * \param[in] an_xml_node Root of XML to dump
+ * \param[in] xml  Root of XML to dump
  *
  * \return Newly allocated buffer containing dumped XML
  */
-static char *
-dump_xml_for_digest(xmlNode * an_xml_node)
+static GString *
+dump_xml_for_digest(xmlNodePtr xml)
 {
-    char *buffer = NULL;
-    int offset = 0, max = 0;
+    GString *buffer = g_string_sized_new(128);
 
     /* for compatibility with the old result which is used for v1 digests */
-    pcmk__buffer_add_char(&buffer, &offset, &max, ' ');
-    pcmk__xml2text(an_xml_node, 0, &buffer, &offset, &max, 0);
-    pcmk__buffer_add_char(&buffer, &offset, &max, '\n');
+
+    g_string_append_c(buffer, ' ');
+    pcmk__xml2text(xml, 0, buffer, 0);
```

Comment:
Coverity thinks we leak memory here. Pretty sure it's wrong. Step 1.7 could never happen since `dump_xml_for_digest()` initializes `buffer`, so it won't be null. If it is somehow null, we've got bigger problems than a memory leak.
```
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/digest.c:41:5:
  1. alloc_fn: Storage is returned from allocation function "pcmk__xml2text".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:1983:5:
...
  1.7. path: Condition "buffer == NULL", taking true branch.
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:2025:17:
  1.8. alloc_fn: Storage is returned from allocation function "g_string_sized_new".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:2025:17:
  1.9. assign: Assigning: "buffer" = "g_string_sized_new(128UL)".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:2027:13:
  1.10. noescape: Resource "buffer" is not freed or pointed-to in function "g_string_append".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:2042:9:
  1.11. return_alloc: Returning allocated memory "buffer".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/digest.c:41:5:
  2. leaked_storage: Ignoring storage allocated by "pcmk__xml2text(xml, 0, buffer, 0)" leaks it.
```

There are four similar Coverity errors for `insert_prefix()`. I'll try assigning the return value to `buffer` every time and see if that makes Coverity happy. It's a no-op when `buffer` starts out non-null.

If that doesn't work, we could make these functions take a `GString **` argument instead of a `GString *` argument and return `void`. Or maybe even if it does work.

Reply:
I went ahead and made them accept `GString **` and return `void`. Hopefully that resolves it. It's more aligned with the usual practice in Pacemaker anyway.